### PR TITLE
Blog MVP + updated streamfield_provider.py so it includes more streamfields

### LINKED ISSUE #3029: Blog MVP
Create a Blog MVP for staff to use.

- [x] Design has been finalized
- [x] Link to original design ticket:  #2911
- [x] Link to related ticket: #3030 (Post MVP blog)
- [x] Link to Redpen: https://redpen.io/p/bha12cf7387de4f7e7
- [x] Design changes been merged into the master Sketch file
- [x] Desktop and mobile mockups are included
- [x] Hover state for elements are designed

## Requirements Checklist
- [ ] Title field
- [ ] Author field
- [ ] Stream field to include
- paragraph component (but that uses the .body-large style instead of the .body style)
- Image
- Image text
- Image text mini 
- Video
- Link button
- Spacer
- Quote
- [ ] always Narrow width 
- [ ] always Zen mode
- [ ] always has IH cta at bottom

## Desktop:
MVP-blog-detail@2x

## Tablet:
MVP-tablet

## Mobile:
MVP-mobile

### PR DESCRIPTION
**Pull request by mmmavis**
Closes #3029 + added 'tags' field to promote tab 

## Checklist

**Changes in Models:**
- [x] Did I update or add new fake data?
- [x] Did I squash my migration?

---

## To create a blog post

Go to page -> homepage -> 'blog' mini site namespace -> create 'blog' page as child page

### CODE REVIEW

File: network-api/networkapi/utility/faker/streamfield_provider.py [hunk 1]
```diff
@@ -113,7 +170,11 @@ def streamfield(self, fields=None):
             'image': generate_image_field,
             'spacer': generate_spacer_field,
             'image_text2': generate_image_text2_field,
-            'quote': generate_quote_field
+            'quote': generate_quote_field,
```
Comment (mmmavis) [inline network-api/networkapi/utility/faker/streamfield_provider.py]: Should I arrange this list alphabetically?

File: network-api/networkapi/wagtailpages/factory.py [hunk 2]
```diff
@@ -15,6 +15,8 @@
 from networkapi.utility.faker import StreamfieldProvider
 
 streamfield_fields = ['header', 'paragraph', 'image', 'spacer', 'image_text2', 'quote']
+blog_body_streamfield_fields = ['paragraph', 'image', 'image_text', 'image_text_mini',
```
Comment (mmmavis) [inline network-api/networkapi/wagtailpages/factory.py]: These are the only streamfield fields we wanna make available for blog post.

File: network-api/networkapi/wagtailpages/migrations/0068_blogpage.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+                ('pub_date', models.DateTimeField(auto_now_add=True)),
+                ('search_image', models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.SET_NULL, related_name='+', to='wagtailimages.Image', verbose_name='Search image')),
```
Comment (mmmavis) [inline network-api/networkapi/wagtailpages/migrations/0068_blogpage.py]: I'm not sure why there's this `search_image` line???

File: network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/ih_cta.html [hunk 4]
```diff
@@ -1,7 +1,5 @@
-<div class="my-5">
```
Comment (mmmavis) [inline network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/ih_cta.html]: Shouldn't need two wrapper divs here.

File: source/sass/wagtail.scss [hunk 5]
```diff
@@ -69,6 +69,14 @@
       max-height: 800px;
     }
   }
+
+  .blog-body {
```
Comment (mmmavis) [inline source/sass/wagtail.scss]: We can do this in SCSS here or modify every non-heading HTML tags in `blog-body` so they are all assigned with the `body-large` class.

e.g., 

```html
<p class="body-large"></p>
<ul class="body-large"></ul>
```

File: network-api/networkapi/wagtailpages/factory.py [hunk 6]
```diff
@@ -199,6 +201,22 @@ class Meta:
...
+        model = networkapi.wagtailpages.models.BlogPage
+        exclude = (
+            'title_text',
+        )
+
+    title = LazyAttribute(lambda o: o.title_text.rstrip('.'))
+    author = Faker('name')
+    body = Faker('streamfield', fields=blog_body_streamfield_fields)
+
+    # Lazy Values
```
Comment (mmmavis) [inline network-api/networkapi/wagtailpages/factory.py]: Copied this comment `# Lazy Values` from an existing section in this file. Not sure why it's called "lazy values" though

File: network-api/networkapi/wagtailpages/models.py [hunk 7]
```diff
@@ -15,6 +15,8 @@
 from wagtail.core.models import Orderable as WagtailOrderable
 from wagtail.images.models import Image
 from modelcluster.fields import ParentalKey
+from modelcluster.contrib.taggit import ClusterTaggableManager
```
Comment (mmmavis) [inline network-api/networkapi/wagtailpages/models.py]: tag related code is basically copied from Wagtail doc: http://docs.wagtail.io/en/v2.0/reference/pages/model_recipes.html#tagging

File: network-api/networkapi/utility/faker/streamfield_provider.py [hunk 8]
```diff
@@ -55,6 +58,36 @@ def generate_image_field():
     })
 
 
+def generate_image_text_field():
+    image_id = choice(Image.objects.all()).id
+    image_text = fake.paragraph(nb_sentences=1, variable_nb_sentences=False)
+    url = fake.url(schemes=['https'])
+    alt_text = ' '.join(fake.words(nb=5))
+    top_divider = fake.boolean()
+    bottom_divider = top_divider
```
Comment (mmmavis) [inline network-api/networkapi/utility/faker/streamfield_provider.py]: `bottom_divider = top_divider` as realistically speaking top and bottom dividers exist as pair.

**Comment by mmmavis** [comment]
@beccaklam if the video is not loading it's probably because the video your linked to doesn't whitelist us (the review app in this case). That's why it's not showing.

Regarding design-y related questions - do you mind chatting with other designers and let me know what the team thinks? There are ways we can do to avoid use `.body-large` in the Image Text or Mini Image Text component. So it really depends on what the team decides! 👍

**Comment by beccaklam** [comment]
> @beccaklam if the video is not loading it's probably because the video your linked to doesn't whitelist us (the review app in this case). That's why it's not showing.
> 
> Regarding design-y related questions - do you mind chatting with other designers and let me know what the team thinks? There are ways we can do to avoid use `.body-large` in the Image Text or Mini Image Text component. So it really depends on what the team decides! 👍

@mmmavis  Alright! So the design team has decided: 
1. Can we bump up line height for .body-large to 1.5? This should work across the site for desktop, mobile and tablet.
2. Can we use .body instead of .body-large for any of the other components (Image Text, Mini Image Text, Captions for videos, images, etc) besides the paragraph component.

**Comment by mmmavis** [comment]
@beccaklam PR updated!

File: network-api/networkapi/wagtailpages/templates/wagtailpages/blog_page.html [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+  {% else %}
+    col-12
+  {% endif %}"
+>
+  <h1 class="h1-heading">{{ page.title }}</h1>
+  <p class="h6-heading mb-5">By {{ page.author }} | {% if page.first_published_at %}{{ page.first_published_at|date:"F j, Y" }}{% else %}not published yet{% endif %}</p>
+  <div class="blog-body">
+    {{ page.body }}
+  </div>
+  {% include "./fragments/ih_cta.html" with wrapper_class="blog-ih-cta-wrapper" %}
```
Comment (mmmavis) [inline network-api/networkapi/wagtailpages/templates/wagtailpages/blog_page.html]: this isn't a streamfield so maybe I should move it out of the `<div class="cms` wrapper

**Comment by mmmavis** [comment]
@Pomax this is ready for review!
@alanmoo can you check out the review app and see if it meets the specs?

**Comment by alanmoo** [comment]
Functionally this looks great! I haven't dug through the code though.

File: network-api/networkapi/wagtailpages/templates/wagtailpages/blog_page.html [hunk 10]
```diff
@@ -0,0 +1,24 @@
+{% extends "./modular_page.html" %}
+
+{% load wagtailcore_tags mini_site_tags %}
+
+{% block pre_footer_classes %}{% if singleton_page %}col-12 col-lg-10 offset-lg-1{% else %}col-md-12{% endif %}{% endblock %}
```
Comment (alanmoo) [inline network-api/networkapi/wagtailpages/templates/wagtailpages/blog_page.html]: Is the `singleton_page` check necessary here?

File: network-api/networkapi/wagtailpages/templates/wagtailpages/blog_page.html [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+{% block subcontent %}
+<div class="py-4 py-md-5
+  {% if page.narrowed_page_content == True %}
```
Comment (alanmoo) [inline network-api/networkapi/wagtailpages/templates/wagtailpages/blog_page.html]: The page is always narrow, why have this check? I believe/hope the plan is to move away from the narrowed page option in general and be more opinionated.

File: network-api/networkapi/wagtailpages/templates/wagtailpages/campaign_page.html [hunk 12]
```diff
@@ -85,7 +85,7 @@ <h1>{{ block.value }}</h1>
           col-12
         {% endif %}"
       >
-        {% include "./fragments/ih_cta.html" %}
+        {% include "./fragments/ih_cta.html" with wrapper_class="my-5" %}
```
Comment (alanmoo) [inline network-api/networkapi/wagtailpages/templates/wagtailpages/campaign_page.html]: At first glance this seems outside the scope of this PR- what's going on?

File: source/sass/wagtail.scss (hunk 5 above)
Comment (alanmoo) [reply source/sass/wagtail.scss]: Here sounds nice

File: network-api/networkapi/utility/faker/streamfield_provider.py (hunk 1 above)
Comment (Pomax) [reply network-api/networkapi/utility/faker/streamfield_provider.py]: the CMS presents them in the order that the code specifies - I don't think we ever really thought about what the best ordering is, so I don't think anyone would object. Pinging @kristinashu just to make sure

File: network-api/networkapi/wagtailpages/factory.py (hunk 6 above)
Comment (Pomax) [reply network-api/networkapi/wagtailpages/factory.py]: The "lazy" refers to the fact that you're not assigning data, you're assigning a generator that will only generate real data when something tries to use it.

The code for that can be seen [here](https://github.com/mozilla/foundation.mozilla.org/blob/f7a67b3cd4db58f8be59aae4aea3b7e21bc19506/network-api/networkapi/wagtailpages/factory.py#L23-L25)

File: network-api/networkapi/utility/faker/streamfield_provider.py (hunk 8 above)
Comment (Pomax) [reply network-api/networkapi/utility/faker/streamfield_provider.py]: Seems reasonable, although as long as the seed is fixed, the fact that they're different values is less important than the fact that they're "the same value each time"

File: network-api/networkapi/utility/faker/streamfield_provider.py (hunk 1 above)
Comment (kristinashu) [reply network-api/networkapi/utility/faker/streamfield_provider.py]: Sorry which order are you talking about it? If it's about the quote component there's actually a bug that multiple quotes don't show up https://github.com/mozilla/foundation.mozilla.org/issues/2845

File: network-api/networkapi/wagtailpages/models.py [hunk 13]
```diff
@@ -517,6 +520,63 @@ class NewsPage(PrimaryPage):
     template = 'wagtailpages/static/news_page.html'
 
 
+class BlogPageTag(TaggedItemBase):
+    content_object = ParentalKey('wagtailpages.BlogPage', on_delete=models.CASCADE, related_name='tagged_items')
```
Comment (Pomax) [inline network-api/networkapi/wagtailpages/models.py]: Is `CASCADE` safe? E.g. are BlogPageTags unique to a single BlogPage? If not, and they're shared, then CASCADE is probably the wrong delete policy

File: network-api/networkapi/wagtailpages/models.py [hunk 14]
```diff
@@ -517,6 +520,63 @@ class NewsPage(PrimaryPage):
...
+    tags = ClusterTaggableManager(through=BlogPageTag, blank=True)
+
+    promote_panels = Page.promote_panels + [
+        FieldPanel('tags'),
+    ]
+
+    # Database fields
+
+    zen_nav = True
+    narrowed_page_content = True
```
Comment (Pomax) [inline network-api/networkapi/wagtailpages/models.py]: Are these flags still relevant?

File: network-api/networkapi/wagtailpages/templates/wagtailpages/blog_page.html (hunk 10 above)
Comment (Pomax) [reply network-api/networkapi/wagtailpages/templates/wagtailpages/blog_page.html]: To rephrase that question, can blog pages ever have children? If not, then the template can be simplified.

File: network-api/networkapi/wagtailpages/templates/wagtailpages/blog_page.html (hunk 11 above)
Comment (Pomax) [reply network-api/networkapi/wagtailpages/templates/wagtailpages/blog_page.html]: Related to that, if the page content is always narrow, then we probably don't need the model to specify that value at all. The template can simply be written to match the fact that blog content is always narrow (with the same comment for zen mode)

File: network-api/networkapi/wagtailpages/templates/wagtailpages/campaign_page.html (hunk 12 above)
Comment (Pomax) [reply network-api/networkapi/wagtailpages/templates/wagtailpages/campaign_page.html]: Agreed, let's make this a tiny follow-up specific to adding the correct wrapper class for campaign pages.

File: network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/ih_cta.html (hunk 4 above)
Comment (Pomax) [reply network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/ih_cta.html]: We shouldn't slip that into this PR - let's leave it doubled here, and file a separate PR for "undoubling" this code

File: network-api/networkapi/wagtailpages/templates/wagtailpages/opportunity_page.html [hunk 15]
```diff
@@ -35,6 +35,6 @@ <h1>{{ block.value }}</h1>
         {% endif %}
       {% endfor %}
     </div>
-    {% include "./fragments/ih_cta.html" %}
+    {% include "./fragments/ih_cta.html" with wrapper_class="my-5" %}
```
Comment (Pomax) [inline network-api/networkapi/wagtailpages/templates/wagtailpages/opportunity_page.html]: out of scope for this PR (see other comments)

File: network-api/networkapi/utility/faker/streamfield_provider.py (hunk 1 above)
Comment (kristinashu) [reply network-api/networkapi/utility/faker/streamfield_provider.py]: Oh you mean this? For consistency in finding things it would be nice if they were in the same order as "opportunity" pages (minus the ones we chose not to included):
image

File: source/sass/wagtail.scss (hunk 5 above)
Comment (mmmavis) [reply source/sass/wagtail.scss]: What we have here should be sufficient enough.
```scss
.blog-body {
    .block-paragraph { // classname for Wagtail 'paragraph' component
      p,
      ul,
      ol,
      a {
        @extend .body-large;
      }
    }
}

```

File: network-api/networkapi/wagtailpages/templates/wagtailpages/blog_page.html (hunk 10 above)
Comment (mmmavis) [reply network-api/networkapi/wagtailpages/templates/wagtailpages/blog_page.html]: @alanmoo ^ ?

File: network-api/networkapi/wagtailpages/templates/wagtailpages/campaign_page.html (hunk 12 above)
Comment (mmmavis) [reply network-api/networkapi/wagtailpages/templates/wagtailpages/campaign_page.html]: The `ih_cta` section came with default `my-5` margin. I had to make these code changes here to make the `ih_cta` section on blog page style properly (`95px` as top margin and with no bottom margin). Bootstrap classes come with `!important` in the rules so I can't override the margins unless I include `!important` in the SCSS code I'm writing for blog post page's `ih_cta`. I'd like to avoid using `!important` to keep our SCSS code sane.

File: network-api/networkapi/wagtailpages/templates/wagtailpages/opportunity_page.html (hunk 15 above)
Comment (mmmavis) [reply network-api/networkapi/wagtailpages/templates/wagtailpages/opportunity_page.html]: see reasoning in https://github.com/mozilla/foundation.mozilla.org/pull/3066#discussion_r281411196

File: network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/ih_cta.html (hunk 4 above)
Comment (mmmavis) [reply network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/ih_cta.html]: see reasoning in https://github.com/mozilla/foundation.mozilla.org/pull/3066#discussion_r281411196

File: network-api/networkapi/wagtailpages/models.py (hunk 13 above)
Comment (mmmavis) [reply network-api/networkapi/wagtailpages/models.py]: This code snippet is directly copied from http://docs.wagtail.io/en/v2.0/reference/pages/model_recipes.html#tagging . @Pomax @alanmoo can you provide code suggestion here as I don't know what appropriate value to put here instead.

File: network-api/networkapi/wagtailpages/templates/wagtailpages/campaign_page.html (hunk 12 above)
Comment (mmmavis) [reply network-api/networkapi/wagtailpages/templates/wagtailpages/campaign_page.html]: If we are okay with letting `ih_cta` use the default style on blog page for now then I'll revert my changes and fix styling as follow-up.

File: network-api/networkapi/wagtailpages/models.py (hunk 14 above)
Comment (mmmavis) [reply network-api/networkapi/wagtailpages/models.py]: I'm removing `narrowed_page_content`. Not sure about `zen_nav` as [primary_nav.html](https://github.com/mozilla/foundation.mozilla.org/blob/master/network-api/networkapi/templates/partials/primary_nav.html) checks `page.zen_nav` in multiple places in the file. Not sure how I can balance "redundant database meta" vs "messy template code". Suggestions?

**Comment by alanmoo** [comment]
Unrelated to the above conversations: Is the font styling in the admin representative of the font styling in the published page? I know that's been an issue in the past, where it's unclear quite what the final page will look like.

File: network-api/networkapi/wagtailpages/templates/wagtailpages/blog_page.html (hunk 10 above)
Comment (alanmoo) [reply network-api/networkapi/wagtailpages/templates/wagtailpages/blog_page.html]: I don't believe they should be able to have children. Let's make sure the option doesn't exist (something like `subpage types` needs to get set somewhere?) and simplify.

File: network-api/networkapi/wagtailpages/templates/wagtailpages/campaign_page.html (hunk 12 above)
Comment (alanmoo) [reply network-api/networkapi/wagtailpages/templates/wagtailpages/campaign_page.html]: Yeah, I'm with Pomax- I think this needs a little more attention and I'm not convinced this is the right approach for it. Back it out just to reduce the complexity of this PR a bit.

File: network-api/networkapi/wagtailpages/models.py (hunk 13 above)
Comment (alanmoo) [reply network-api/networkapi/wagtailpages/models.py]: Playing with the review app, this seems to work as expected- adding the same tag to multiple posts, then removing it from one leaves it on the others.

File: network-api/networkapi/utility/faker/streamfield_provider.py (hunk 1 above)
Comment (mmmavis) [reply network-api/networkapi/utility/faker/streamfield_provider.py]: Artboard 1

Alright. Looks like we are good here!

File: network-api/networkapi/wagtailpages/templates/wagtailpages/campaign_page.html (hunk 12 above)
Comment (mmmavis) [reply network-api/networkapi/wagtailpages/templates/wagtailpages/campaign_page.html]: follow-up ticket filed: https://github.com/mozilla/foundation.mozilla.org/issues/3111

File: network-api/networkapi/wagtailpages/templates/wagtailpages/blog_page.html (hunk 10 above)
Comment (mmmavis) [reply network-api/networkapi/wagtailpages/templates/wagtailpages/blog_page.html]: https://github.com/mozilla/foundation.mozilla.org/search?q=pre_footer_classes&unscoped_q=pre_footer_classes

image

It seems like `pre_footer_classes` block isn't doing anything. I'll update `blog_page.html` and file a ticket to remove `{% block pre_footer_classes %}` for good

File: network-api/networkapi/management/commands/load_fake_data.py [hunk 16]
```diff
@@ -369,6 +371,19 @@ def reseed():
 
         reseed()
 
+        try:
+            blog_namespace = WagtailPage.objects.get(title='blog')
+            print('blog namespace exists')
+        except WagtailPage.DoesNotExist:
+            print('Generating a blog namespace')
+            blog_namespace = MiniSiteNameSpaceFactory.create(parent=home_page, title='blog', live=False)
```
[the remark this answers is not in the document]
Comment (mmmavis) [reply network-api/networkapi/management/commands/load_fake_data.py]: PR updated! @Pomax can you verify and 'resolve conversation' if my code changes look good?

**Comment by mmmavis** [comment]
@Pomax @alanmoo PR is ready for review again (I just need to fix one linting error)! Hopefully I've addressed everything mentioned. I left some comment threads open as I didn't know if I really resolved those issues in my PR.

**Comment by alanmoo** [comment]
Filed #3155 for Percy testing

**Comment by mmmavis** [comment]
🎉 🎉 🎉 🎉 🎉
